### PR TITLE
Soundcloud Channel

### PR DESCRIPTION
**Pull request by byronwasti**
Finally fixed it to work with the rest of the system. Implements the soundcloud channel, along with a little endian steganography module. The lendianSteganography, as of right now, only supports encoding the data in randomly generated data but it should be very easy to expand if we want other functionality from it.

### CODE REVIEW

**Comment by byronwasti** [comment]
There is an issue with de-steganography. I have to figure out how to make the system work with the current set up.

**Comment by byronwasti** [comment]
Temporary solution to work with data[0] input to decoder. We should look into figuring out why we do that/if it is necessary.

**Comment by DakotaNelson** [comment]
Can you add any dependencies to the readme? It looks like the `soundcloud` package is a pip install.

File: channels/test.txt [hunk 1]
```diff
@@ -0,0 +1 @@
... is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.This is some test text file.
```
Comment (DakotaNelson) [inline channels/test.txt]: Does this need to be committed?

File: encoders/lendianSteganography.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+# Lendian stegonography system
```
Comment (DakotaNelson) [inline encoders/lendianSteganography.py:1]: Could you add some documentation noting that this only works for sound?

As it stands, I would assume it works on images.

File: encoders/lendianSteganography.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            tmp_data = randomSound()
+            if i == 0:
+                tmp_data = tmp_data & ~1
+            else:
+                tmp_data = tmp_data | 1
+
+            encoded_data.append(str(tmp_data))
+
+    else:
+        #TODO: Is this how we are handling errors?
```
Comment (DakotaNelson) [inline encoders/lendianSteganography.py]: Try raising an error that's more descriptive; remember, the user won't know which module had an issue, they'll just run a loooong command and then get this non-specific error message.

File: screep [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python2
```
Comment (DakotaNelson) [inline screep:1]: Good catch, thanks. :+1:

File: encoders/lendianSteganography.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    # in case encoding dataset that is bigger than
+    # data size
+    for i in range(len(data)/bit_size):
+        singleChar = [int(i) for i in data[i*8:8+i*8]]
+        decoded_data.append(decodeChar(singleChar))
+    
+    return ''.join(decoded_data)
+
+if __name__ == "__main__":
+    data_set = "Hello this is a large test of the de-stegranography system"
```
Comment (DakotaNelson) [inline encoders/lendianSteganography.py:85]: Huh, this is a cool idea. I like it.

We have the `echo` subcommand for stuff like this now, but I also like testing it in isolation like this.

File: channels/soundcloudChannel.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    wf.setsampwidth(2)
+    wf.writeframes(b''.join(frames))
+    wf.close()
+
+    print("Done creating sound file")
+    track = client.post('/tracks', track={
+        'title': params['song_name'],
+        'sharing':'public',
+        'asset_data': open('output.wav','rb'),
+        'tag_list':'tag1 \"hip hop\"',
```
Comment (DakotaNelson) [inline channels/soundcloudChannel.py:47]: Makes sense.

File: encoders/lendianSteganography.py (hunk 2 above)
Comment (byronwasti) [inline encoders/lendianSteganography.py:1]: It does work on images, technically. Right now it can only encode text into random data that it generates.

**Comment by byronwasti** [comment]
Good stuff, I think I fixed everything that was out of place/not clear.

File: channels/soundcloudChannel.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        client_secret=params['secret'],
+        username=params['username'],
+        password=params['password']
+    )
+
+    frames = []
+
+    for i in data:
+        frames.append(i)
+        frames.append(',')
```
Comment (DakotaNelson) [inline channels/soundcloudChannel.py:33]: Does there need to be any sanity checking on the frames? If I just ran some data through b64 and passed it to this, would it work?

If not, just do some quick sanity checking and raise an [appropriate exception](https://docs.python.org/2/library/exceptions.html) if something's wrong. (Better yet, make it so it'll be compatible - if that's something that can be done.)

I just want to make sure that module linkages (i.e. _this_ encoder must be run before _that_ module) are as few as we can make them, and that they're really obvious when they happen.

**Comment by DakotaNelson** [comment]
Just added another question, but this looks fantastic - feel free to merge it in once the above has been addressed. I'm incredibly excited to release this - awesome work, @byronwasti! :smiley:

File: channels/soundcloudChannel.py (hunk 7 above)
Comment (byronwasti) [inline channels/soundcloudChannel.py:33]: There isn't a need for sanity checking. Just tried it to confirm and it worked.

**Pull request merged**